### ISSUE TITLE
GA Alternator Streams

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#16367]
Three years ago, we added (https://github.com/scylladb/scylladb/pull/6694) support for the DynamoDB Streams feature (https://github.com/scylladb/scylladb/issues/5065). It is based on CDC, which was _experimental_ at the time, so Alternator Streams was also deemed an experimental feature. Later, when CDC became GA (commit 78649c2322c63f3156440d86bdecb1c16ddb2efe), we decided to leave the Alternator Streams experimental, and a special option, `'--experimental-features=alternator-streams`, was added to enable it.

The purpose of this issue is to call for finally GA'ing (i.e. removing the need for an "experimental-features" option) Alternator Streams, and to outline what needs to be done (if anything) before that.

Here is a list of Alternator Streams issues that we either need to fix before GA, or make a decision that a GA can happen despite these issues not being solved (GA doesn't need to mean perfect and usable with every use case - it can have known, documented, problems and limitations).

* #6918 - Alternator Streams does does not properly distinguish the type of change event
* #6929 - Reconsider default alternator_streams_time_window_s = 10 seconds
* #6930 - Alternator Streams thinks that PutItem is a REMOVE+MODIFY
* #6931 - Alternator Streams: minor incompatibilities in GetRecords response fields
* #6939 - Alternator Streams: changing an existing stream's StreamViewType should be forbidden
* #7239 - Alternator Stream should still be usable (for 24 hours) after disabling
* #11523 - Alternator Streams events for expired items (TTL) should be specially flagged
* #13080 - Too many "shards" in Alternator Streams

**Comment by zey1996** [comment]
Maybe we can add some doc, before we resolve this problem.  
I am trying use cql to query cdc_log table. for sync of a alternator table. it's better then alternator stream in some problem, like https://github.com/scylladb/scylladb/issues/13080. 
Maybe I can do some thing?

**Comment by nyh** [comment]
> Maybe we can add some doc, before we resolve this problem.

A doc about what, summarizing all the differences from DynamoDB Streams and pitfalls one must be aware of when migrating a DynamoDB Streams application to Alternator?
We do have a very brief summary of these differences in docs/alternator/compatibility.md, but maybe you were hoping for a longer document?

> I am trying use cql to query cdc_log table. for sync of a alternator table. it's better then alternator stream in some problem, like #13080.

Interesting. Reading the CDC log using CQL will not help some parts of #13080 - namely the part that you really need to read from many different partitions - but will help avoid the biggest problem in #13080 which is that certain existing libraries for DynamoDB Streams handle these many partitions very badly (e.g., reading each of them in a separate thread, with terrible overhead). Obviously, reading the CDC log with CQL means you will not be using any of these broken libraries. Will you be using Scylla's library for CQL CDC or your own implementation? If it's the latter, then you might consider using Alternator Streams, not CQL, in your implementation.

> Maybe I can do some thing?

What do you propose?

**Comment by zey1996** [comment]
1. yes, but I think it's not enough. example:   
- why we should use time winodw10s. https://github.com/scylladb/scylladb/issues/6929. In fact scylla-cdc-go use time window is 30s. I still don't know what the appropriate value to set is. And why we must need it.  
- https://github.com/scylladb/scylladb/issues/6930. Should we explain event correspondences and how users should handle them. for example, I set `StreamViewType="KEYS_ONLY"`. If I putItem by alternator. I got three record in cdc log table(delete, insterm update). This is a little confusing.  
- 
2. In fact, I think cql stream is better than alternator. And yes, I use scylla-cdc-go.  In this test, I create a cluster of 3 node 1 dc, and smp=2. this is why:  
first, the number of stream is 1536(256 * 3 * 2). If I use alternator I should 1536 query to check all stream. But if I use scylla-cdc-go, I only need 768 query. This makes a significant difference in many core nodes.
second. In my test, scylla version is 4.6. scylla-cdc-go is better than alternator in scylla load level.  I think reason is I can't use shard-aware connect to alternator. 
I hope I can increase scylla's load as little as possible and get data change event as soon as possible.

3. we should remind that alternator is different from dynamodb, And explain why. And I think we can give an example, For instructions how to use alternator like dynamodb. maybe we can add a example, it can conver cdc log event to dynamodb stream event.

**Comment by nyh** [comment]
@zey1996 you raised some interesting points. Definitely a separate document about the remaining Alternator Streams pitfalls can be useful,  or alternatively we could expand the single paragraph about Streams in docs/alternator/compatibility.md to a longer section.

I don't know what to think about using CQL (and a library like scylla-cdc-go) to query the stream. I agree there might be performance advantages, but it's possible that the same can be said also for normal (non-stream) requests, and still we don't want to support that - an Alternator table should be accessed with the Alternator (DynamoDB) API, not through CQL even if the latter can have some performance advantage. If you do access the Alternator table, or stream, via CQL, you'll also need to understand the way that DynamoDB data is encoded in CQL - special column names like ":attrs" for non-key attributes, how different types (strings, numbers, lists, etc.) are encoded, and so on. So far, we didn't document this because we didn't really want users to rely on these encodings.

**Comment by zey1996** [comment]
@nyh I agree that querying streams using CQL is difficult. But there are really big differences between the way alternator and cql query streams. This is mainly because alternator queries one streamID per request, but scylla-cdc-go queries one vnode per request(many streamID), which greatly reduces the number of requests.

**Comment by ScyllaPiotr** [comment]
I created [a separate document](https://docs.google.com/document/d/1vEI5qEPWjsSOttsGTjvkROJFOOsp_A0BNbu9pM3S_tY/) with the proposed classification of the tasks (needed / not needed) and to enable an organised (threaded) discussion.

**Comment by ScyllaPiotr** [comment]
@nyh please reference in the 1st msg the new one too: #24598

**Comment by ScyllaPiotr** [comment]
The discussion has been concluded with a version accepted on the Core Arch meeting. The (internal) document is [here](https://scylladb.atlassian.net/wiki/spaces/RND/pages/98828836/Alternator+Streams+v1+GA+prerequisites).

**Comment by nyh** [comment]
> The discussion has been concluded with a version accepted on the Core Arch meeting. The (internal) document is [here](https://scylladb.atlassian.net/wiki/spaces/RND/pages/98828836/Alternator+Streams+v1+GA+prerequisites).

I updated the list on the top according to what you wrote in that document.

**Comment by ScyllaPiotr** [comment]
Referenced in Jira for mgmt visibility: https://scylladb.atlassian.net/browse/SCYLLADB-461
